tests/test_visitor.py

```diff
@@ -1255,3 +1255,49 @@ def test_attach_scopes_associates(frontend):
     assert var_map['var%foo'].scope is associates[0]
     assert var_map['var%foo'].parent.scope is associates[0]
     assert var_map['var%foo'].parent is var_map['var']
+
+
+@pytest.mark.parametrize('invalidate_source', [True, False])
+@pytest.mark.parametrize('replacement', ['body', 'self', 'self_tuple', 'duplicate'])
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_transformer_duplicate_node_tuple_injection(frontend, invalidate_source, replacement):
+    """Test for #41, where identical nodes in a tuple have not been
+    correctly handled in the tuple injection mechanism."""
+    fcode_kernel = """
+SUBROUTINE compute_column(start, end, nlon, nz, q)
+    INTEGER, INTENT(IN) :: start, end
+    INTEGER, INTENT(IN) :: nlon, nz
+    REAL, INTENT(INOUT) :: q(nlon,nz)
+    INTEGER :: jl
+    DO JL = START, END
+        Q(JL, NZ) = Q(JL, NZ) * 0.5
+    END DO
+    DO JL = START, END
+        Q(JL, NZ) = Q(JL, NZ) * 0.5
+    END DO
+END SUBROUTINE compute_column
+"""
+    kernel = Subroutine.from_source(fcode_kernel, frontend=frontend)
+    kernel.body = SubstituteExpressions({}, invalidate_source=invalidate_source).visit(kernel.body)
+    loops = FindNodes(Loop).visit(kernel.body)
+    if replacement == 'body':
+        # Replace loop by its body
+        mapper = {l: l.body for l in loops}
+    elif replacement == 'self':
+        # Replace loop by itself
+        mapper = {l: l for l in loops}
+    elif replacement == 'self_tuple':
+        # Replace loop by itself, but wrapped in a loop
+        mapper = {l: (l,) for l in loops}
+    elif replacement == 'duplicate':
+        # Duplicate the loop (will this trigger infinite recursion in tuple injection)?
+        mapper = {l: (l, l) for l in loops}
```

Comment:
Could we add a check here somewhere that the replacement has happened? Something line `len(FindNodes(Loop).visit(routine.body)) == 1/2/3`?